icmp.go

```diff
@@ -0,0 +1,495 @@
+}
+
+func (r *Hop) Fields() log.Fields {
+	return map[string]interface{}{
+		"srchost":    r.Src.Host,
+		"srcip":      r.Src.IP.String(),
+		"dsthost":    r.Dst.Host,
+		"dstip":      r.Dst.IP.String(),
+		"nodehost":   r.Node.Host,
+		"nodeip":     r.Node.IP.String(),
+		"step":       r.Step,
+		"id":         r.ID,
+		"sent":       r.Sent.Format(time.RFC3339Nano),
+		"received":   r.Received.Format(time.RFC3339Nano),
+		"elapsed":    r.Elapsed.Seconds(),
+		"generation": r.Generation,
+	}
+}
+
+func (c *Continuous) extractMessage(p []byte, now time.Time) (*Hop, error) {
+	// get the reply IPv4 header. That will have the node address
+	replyHeader, err := icmp.ParseIPv4Header(p)
+	if err != nil {
+		return nil, err
+	}
+	// now, extract the ICMP message
+	msg, err := icmp.ParseMessage(syscall.IPPROTO_ICMP, p[replyHeader.Len:])
+	if err != nil {
+		return nil, err
+	}
+	icmpType := int(p[replyHeader.Len])
+	var data []byte
+	if te, ok := msg.Body.(*icmp.TimeExceeded); ok {
+		data = te.Data
+	} else if du, ok := msg.Body.(*icmp.DstUnreach); ok {
+		data = du.Data
+	} else {
+		return nil, fmt.Errorf("Unknown message type: %v", msg.Type)
+	}
+	// data should now have the IP header of the original message plus at least
+	// 8 bytes of the original message (which is, at least, the UDP header)
+	srcHeader, err := icmp.ParseIPv4Header(data)
+	if err != nil {
+		return nil, err
+	}
+	udpHeader := data[srcHeader.Len:]
+	if len(udpHeader) < 8 {
+		return nil, fmt.Errorf("source udp header too short: %d", len(udpHeader))
+	}
+	srcPort := binary.BigEndian.Uint16(udpHeader[0:2])
+	dstPort := binary.BigEndian.Uint16(udpHeader[2:4])
+	var name string
+	names, _ := net.LookupAddr(replyHeader.Src.String())
+	if len(names) > 0 {
+		name = names[0]
+	} else {
+		name = replyHeader.Src.String()
+	}
+	id := c.packetID(srcHeader.ID, srcHeader.Dst, int(srcPort), int(dstPort))
+	return &Hop{
+		Src: Addr{
+			IP: srcHeader.Src,
+		},
+		Dst: Addr{
+			IP: srcHeader.Dst,
+		},
+		Node: Addr{
+			Host: name,
+			IP:   replyHeader.Src,
+		},
+		Received: now,
+		ID:       id,
+		icmpType: icmpType,
+	}, nil
+}
+
+func (c *Continuous) readICMP(ctx context.Context, hops chan *Hop) error {
+	defer close(hops)
+	localAddr, err := findAddress()
+	if err != nil {
+		return err
+	}
+	// set up the listening socket for incoming ICMP packets
+	recvSocket, err := syscall.Socket(syscall.AF_INET, syscall.SOCK_RAW, syscall.IPPROTO_ICMP)
+	if err != nil {
+		// this is fatal, because we need to be able to read the responses
+		return err
+	}
+	defer syscall.Close(recvSocket)
+	// TODO: do we need a receive timeout if we're running continuously?
+	// we do, if we want ctx.Done() to exit the recvfrom loop, otherwise it blocks forever
+	// once we stop sending requests
+	// tv := syscall.NsecToTimeval(int64(c.Timeout))
+	// err = syscall.SetsockoptTimeval(recvSocket, syscall.SOL_SOCKET, syscall.SO_RCVTIMEO, &tv)
+	// if err != nil {
```

Comment:
Confused - is this commented out code here for a reason, or related to the todo? This fatal error is different from non-commented out one above?

Reply:
It's there because I can't make up my mind whether we really want/need a timeout. I added some decorations to group the comment with the code it references.